### PR TITLE
SOLR-15489: Implement OFFSET & FETCH for LIMIT queries

### PR DESCRIPTION
**Pull request by thelabdude**
https://issues.apache.org/jira/browse/SOLR-15489

# Description

Add support for OFFSET & FETCH for Parallel SQL queries; only supports limit queries!
```
SELECT id FROM $ALIAS ORDER BY id DESC OFFSET 5 FETCH NEXT 5 ROWS ONLY
```

# Solution

Enhanced the LimitStream to support an offset. If offset is provided, then we wrap the CloudSolrStream with a SortStream and then use the LimitStream with offset.

# Tests

Added unit test to verify the correct rows returned based on offset / limit used.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [x] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/handler/sql/SolrTable.java [hunk 1]
```diff
@@ -275,25 +277,43 @@ private TupleStream handleSelect(String zk,
...
     if (limit != null) {
-      params.add(CommonParams.ROWS, limit);
-      return new LimitStream(new CloudSolrStream(zk, collection, params), Integer.parseInt(limit));
+      int limitInt = Integer.parseInt(limit);
+      // if there's an offset, then we need to fetch offset + limit rows from each shard and then sort accordingly
+      LimitStream limitStream;
+      if (offset != null) {
+        int offsetInt = Integer.parseInt(offset);
+        int rows = limitInt + offsetInt;
+        params.add(CommonParams.START, "0"); // tricky ... we need all rows up to limit + offset
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/handler/sql/SolrTable.java:307]: This makes me nervous... at what point do we need to get back into cursor mark world?

Comment (thelabdude) [reply solr/core/src/java/org/apache/solr/handler/sql/SolrTable.java:307]: How would you map a cursor mark to SQL? Not sure why this makes you nervous as it's not exposing any more risk than people doing:
```
q=*:*&start=1000000000000&rows=10
```
Solr's query interface allows all kinds of dangerous input if people want to mis-use the parameters.

If you have an idea on how to map cursorMark to SQL I'd appreciate hearing the details.

File: solr/core/src/test/org/apache/solr/handler/TestSQLHandler.java [hunk 2]
```diff
@@ -1853,21 +1853,18 @@ public void testParallelTimeSeriesGrouping() throws Exception {
...
-    List<Tuple> tuples = new ArrayList<>();
-    for (; ; ) {
-      Tuple t = tupleStream.read();
-      //log.info(" ... {}", t.fields);
-      if (t.EOF) {
-        break;
-      } else {
-        tuples.add(t);
+    List<Tuple> tuples = new LinkedList<>();
+    try (TupleStream tupleStream = new SolrStream(baseUrl, params)) {
```
Comment (madrob) [inline solr/core/src/test/org/apache/solr/handler/TestSQLHandler.java:1857]: This pattern makes me wish we had an `asStream` method on TupleStream that returned a java Stream. Maybe a future issue.

Comment (thelabdude) [reply solr/core/src/test/org/apache/solr/handler/TestSQLHandler.java:1857]: Maybe? Here I'm fixing a bug where the stream didn't get closed if an exception gets raised, causing tests to hang.

**Review by markrmiller — COMMENTED** [review]
I think it is possible to use CursorMark in some form for this.

Given that pagination is semi accurate (and generally not very valuable :)), I think that leaves some attack path opens.

The CursorMark is really just last values for the sort to be used to indicate the starting offset.

So you could come up with some caching, loose effort type approaches I think.

For deep paging, it seems likely worth some effort. I started to look into it, but the actually value and likely practical is so low.

I'm +1 on starting here.

**Pull request merged**